File: Memo01/app/src/main/java/com/callor/memo/db/ViewModelFactory.java
package com.callor.memo.db;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class ViewModelFactory  implements ViewModelProvider.Factory {


    private final MemoDataSource mDataSource;

    public ViewModelFactory(MemoDataSource mDataSource) {
        this.mDataSource = mDataSource;
    }


    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(MemoViewModel.class)) {
            return (T) new MemoViewModel(mDataSource);
        }
        //noinspection unchecked
        throw new IllegalArgumentException("Unknown ViewModel class");

    }
}

File: MemoKt01/app/src/main/java/com/callor/memo/db/ViewModelFactory.kt
package com.callor.memo.db

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class ViewModelFactory(private val mDataSource: MemoDataSource) : ViewModelProvider.Factory {


    override fun <T : ViewModel> create(modelClass: Class<T>): T {

        if (modelClass.isAssignableFrom(MemoViewModel::class.java)) {
            return MemoViewModel(mDataSource) as T
        }

        throw IllegalArgumentException("Unknown ViewModel class")

    }
}
